### PR TITLE
Introduce wallet provider

### PR DESCRIPTION
**Pull request by AurelienFT**
This PR : 

- Remove the limit to have a provider of private API.
- Use random provider to make an API call so that if we pass a big list from wallet-provider it will switch between nodes often.
- Abstract low level operation creation in 2 mode : One with local secret key, one using wallet provider
- Add a constructor of client that take account of wallet provider as parameters.

This PR is used in an example to test it : https://github.com/massalabs/massa-sc-examples/pull/157

### CODE REVIEW

**Comment by AurelienFT** [comment]
@gregLibert Can we ignore the security concerns as it's not important in our use case ?

File: src/web3/Client.ts [hunk 1]
```diff
@@ -26,15 +33,30 @@ export class Client implements IClient {
...
+    baseAccount?: IAccount | IAccountWalletProvider,
   ) {
     this.publicApiClient = new PublicApiClient(clientConfig);
     this.privateApiClient = new PrivateApiClient(clientConfig);
-    this.walletClient = new WalletClient(
-      clientConfig,
-      this.publicApiClient,
-      baseAccount,
-    );
+    // From: https://stackoverflow.com/a/73162923/10419533
```
Comment (gregLibert) [inline src/web3/Client.ts:40]: Not sure it's linked to what you are doing.
Here your simply defining the base account based on the wanted context. If you don't want to go the direction I suggested above, you could improve readibility by doing something like:
``` typescript
this.walletClient = new WalletClient(
        clientConfig,
        this.publicApiClient,
        provider(baseAccount, this.publicApiClient),
      );
 ```
 And somewhere else define the provider function:
 ```typescript
  private provider(): Account {
  if (!baseAccount) {
    return undefined;
  }

  if (isWeb3Account(baseAccount)) {
    return new Web3Account(baseAccount, this.publicApiClient);
  }
  
  return new WalletProviderAccount(baseAccount);
  }

File: src/web3/Client.ts [hunk 2]
```diff
@@ -26,15 +33,30 @@ export class Client implements IClient {
    */
   public constructor(
     private clientConfig: IClientConfig,
-    baseAccount?: IAccount,
+    baseAccount?: IAccount | IAccountWalletProvider,
```
Comment (gregLibert) [inline src/web3/Client.ts:36]: Why not using a IBaseAccount instead and let people instantiate Client class using meaningful helper such as:
const client =  newBrowserWalletProvider() or const client = newNodeJSProvider() ?
this abviously imply to define the function newBrowserWalletProvider and newNodeJSProvider

**Review by gregLibert — COMMENTED** [review]
To be discussed in live meeting

File: src/web3/BaseClient.ts [hunk 3]
```diff
@@ -185,7 +173,9 @@ export class BaseClient {
       case JSON_RPC_REQUEST_METHOD.ADD_STAKING_PRIVATE_KEYS:
       case JSON_RPC_REQUEST_METHOD.NODE_SIGN_MESSAGE:
       case JSON_RPC_REQUEST_METHOD.NODE_REMOVE_FROM_WHITELIST: {
-        return this.getPrivateProviders()[0]; // TODO: choose the first available private provider ?
+        let providers = this.getPrivateProviders();
```
[the remark this answers is not in the document]
Comment (AurelienFT) [reply src/web3/BaseClient.ts:176]: Yes but it's not mandatory anymore.

File: src/web3/Client.ts [hunk 4]
```diff
@@ -7,7 +7,14 @@ import { SmartContractsClient } from './SmartContractsClient';
 import { IProvider, ProviderType } from '../interfaces/IProvider';
 import { DefaultProviderUrls } from './ClientFactory';
 import { IClient } from '../interfaces/IClient';
-import { IWalletClient } from '../interfaces/IWalletClient';
+import { IAccount as IAccountWalletProvider } from '@massalabs/wallet-provider';
+import { Web3Account } from './accounts/Web3Account';
+import { WalletProviderAccount } from './accounts/WalletProviderAccount';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isWeb3Account(item: any): item is IAccount {
```
[the remark this answers is not in the document]
Comment (AurelienFT) [reply src/web3/Client.ts]: Removed

File: src/web3/Client.ts (hunk 1 above)
Comment (AurelienFT) [reply src/web3/Client.ts]: Changed the behavior to have IBaseAccount

File: src/web3/Client.ts (hunk 2 above)
Comment (AurelienFT) [reply src/web3/Client.ts]: Done

**Review by Ben-Rey — APPROVED** [review]

**Review by gregLibert — APPROVED** [review]

**Pull request merged**